### ISSUE TITLE
4.1.0-rc1 fails to start with ldap backend and master=yes

### ISSUE THREAD

**Issue by sooslaca** [PowerDNS/pdns#5742]
- Program: Authoritative
 - Issue type: Bug report 

### Short description

pdns_server 4.1.0-rc1 failes to start with ldap when master=yes

### Environment

 - Operating system: Docker, Alpine 3.6
 - Software version: 4.1.0-rc1
 - Software source: https://downloads.powerdns.com/releases/pdns-4.1.0-rc1.tar.bz2

### Steps to reproduce

1. in Dockerfile compile pdns 4.1.0-rc1, default flags, search apline apk pdns 4.0.3. same I used, no change
2. Configfile very simple
daemon=no
launch=ldap
master=yes
ldap-host=ldap://ldap.internal/
ldap-basedn=dc=top
ldap-binddn=cn=admin,dc=top
ldap-secret=asdasd123
ldap-method=simple
ldap-timeout=50

### Expected behaviour

when I start pdns_server it's expected to start without issues

### Actual behaviour

Sep 28 12:03:36 Done launching threads, ready to distribute questions
Sep 28 12:03:41 [LdapBackend] Unable to search LDAP directory: Timeout
Sep 28 12:03:41 [LdapBackend] Ldap connection closed
Sep 28 12:03:41 Exiting because communicator thread died with error: LDAP server timeout

### Other information

Exact same configuration works with 4.0.4
When commenting out "master=yes" it works

**Comment by pieterlexis** [comment]
quick note, it looks like we don't start pdns_server with `--master` in the regression tests for ldap

**Comment by Habbie** [comment]
> quick note, it looks like we don't start pdns_server with --master in the regression tests for ldap

good catch! I'll have travis just test that now.

**Comment by mind04** [comment]
timeout = 50 (seconds) and within a few seconds after startup a timeout...

**Comment by pieterlexis** [comment]
when we do pass `--master` to the auth in the regression tests, all [tests succeed](https://travis-ci.org/pieterlexis/pdns/jobs/280848651). Would it be possible for you to test this in a container with a a non-musl based libc (like ubuntu or centos)?

**Comment by sooslaca** [comment]
Fails on Debian container as well if I add master=yes. If I delete then works.

root@fbf7028f3e10:/usr/src/pdns# cat /etc/issue
Debian GNU/Linux 9 \n \l

root@fbf7028f3e10:/# pdns_server
...
Sep 28 15:41:30 PowerDNS Authoritative Server 4.1.0-rc1 (C) 2001-2017 PowerDNS.COM BV
Sep 28 15:41:30 Using 64-bits mode. Built using gcc 6.3.0 20170516 on Sep 28 2017 15:23:24 by root@fbf7028f3e10.
...
Sep 28 15:41:14 Done launching threads, ready to distribute questions
Sep 28 15:41:19 [LdapBackend] Unable to search LDAP directory: Timeout
Sep 28 15:41:19 [LdapBackend] Ldap connection closed
Sep 28 15:41:19 Exiting because communicator thread died with error: LDAP server timeout

**Comment by Habbie** [comment]
> timeout = 50 (seconds) and within a few seconds after startup a timeout...

But exactly 5 seconds which is the default for timeout!

**Comment by sooslaca** [comment]
If it helps, here are the logs from the LDAP container. Search happens once then there are 3 other auth - successful - attempts. 
See timings as well, in 5 seconds it unbinds.

Sep 28 15:41:14 dockerhost docker[32237]: 59cd181a conn=1066 fd=22 ACCEPT from IP=172.18.0.1:45454 (IP=0.0.0.0:389)
Sep 28 15:41:14 dockerhost docker[32237]: 59cd181a conn=1066 op=1 SRCH base="dc=top" scope=2 deref=3 filter="(&(sOARecord=*)(PdnsDomainId=*))"
Sep 28 15:41:14 dockerhost docker[32237]: 59cd181a conn=1066 op=1 SRCH attr=associatedDomain
Sep 28 15:41:14 dockerhost docker[32237]: 59cd181a conn=1066 op=1 SEARCH RESULT tag=101 err=0 nentries=1 text=
Sep 28 15:41:14 dockerhost docker[32237]: 59cd181a conn=1066 op=2 SRCH base="dc=top" scope=2 deref=3 filter="(&(associatedDomain=example.com)(sOARecord=*))"
Sep 28 15:41:14 dockerhost docker[32237]: 59cd181a conn=1066 op=2 SRCH attr=sOARecord PdnsDomainId PdnsDomainNotifiedSerial PdnsDomainLastCheck PdnsDomainMaster PdnsDomainType
Sep 28 15:41:14 dockerhost docker[32237]: 59cd181a conn=1066 op=2 SEARCH RESULT tag=101 err=0 nentries=1 text=
Sep 28 15:41:14 dockerhost docker[32237]: 59cd181a conn=1068 fd=23 ACCEPT from IP=172.18.0.1:45458 (IP=0.0.0.0:389)
Sep 28 15:41:14 dockerhost docker[32237]: 59cd181a conn=1068 op=0 BIND dn="cn=admin,dc=top" method=128
Sep 28 15:41:14 dockerhost docker[32237]: 59cd181a conn=1068 op=0 BIND dn="cn=admin,dc=top" mech=SIMPLE ssf=0
Sep 28 15:41:14 dockerhost docker[32237]: 59cd181a conn=1068 op=0 RESULT tag=97 err=0 text=
Sep 28 15:41:14 dockerhost docker[32237]: 59cd181a conn=1069 fd=24 ACCEPT from IP=172.18.0.1:45464 (IP=0.0.0.0:389)
Sep 28 15:41:14 dockerhost docker[32237]: 59cd181a conn=1069 op=0 BIND dn="cn=admin,dc=top" method=128
Sep 28 15:41:14 dockerhost docker[32237]: 59cd181a conn=1069 op=0 BIND dn="cn=admin,dc=top" mech=SIMPLE ssf=0
Sep 28 15:41:14 dockerhost docker[32237]: 59cd181a conn=1069 op=0 RESULT tag=97 err=0 text=
Sep 28 15:41:14 dockerhost docker[32237]: 59cd181a conn=1070 fd=25 ACCEPT from IP=172.18.0.1:45468 (IP=0.0.0.0:389)
Sep 28 15:41:14 dockerhost docker[32237]: 59cd181a conn=1070 op=0 BIND dn="cn=admin,dc=top" method=128
Sep 28 15:41:14 dockerhost docker[32237]: 59cd181a conn=1070 op=0 BIND dn="cn=admin,dc=top" mech=SIMPLE ssf=0
Sep 28 15:41:14 dockerhost docker[32237]: 59cd181a conn=1070 op=0 RESULT tag=97 err=0 text=
Sep 28 15:41:19 dockerhost docker[32237]: 59cd181f conn=1066 op=3 UNBIND
Sep 28 15:41:19 dockerhost docker[32237]: 59cd181f conn=1066 fd=22 closed
Sep 28 15:41:20 dockerhost docker[32237]: 59cd1820 conn=1070 fd=25 closed (connection lost)
Sep 28 15:41:20 dockerhost docker[32237]: 59cd1820 conn=1069 fd=24 closed (connection lost)
Sep 28 15:41:20 dockerhost docker[32237]: 59cd1820 conn=1067 fd=21 closed (connection lost)
Sep 28 15:41:20 dockerhost docker[32237]: 59cd1820 conn=1068 fd=23 closed (connection lost)

**Comment by aerique** [comment]
@sooslaca I'm not able to reproduce this issue using either the latest pull (efaa74ecc48026f2548a3cefacdea0331bbe7c88) from master or the abovementioned software source.

*However*, I am not running the auth server from Docker. So could you either give me the exact steps you've taken to build the auth server inside your Docker container example or try it outside of Docker?

I've tested against this LDAP docker image: https://github.com/osixia/docker-openldap which I started as follows `docker run --name my-openldap-container osixia/openldap:1.1.9`. I did a `docker inspect «container-id»` to find out which IP address I had to configure in your example configuration.

@mind04 Your comment is a little succinct. Am I correct in assuming you ran into the same issue? If so, could you give me your reproduction steps as well? Was it also with the auth server inside Docker? If so, could it be a Docker network issue?

Thanks!

**Comment by sooslaca** [comment]
I apologise I’m being away from computer until 14th Oct.
I’ll surely give you the Dockerfile after that time for both the ldap and the pdns containers.
Its very simple though, grab an Alpine 3.6 base image and compile ldap and pdns from source. You can grab dependencies and runtime libararies from apk package source (x64).
Docker host is Ubuntu LTS.

Docker run it with - -init and specific network (- -network x)

**Comment by aerique** [comment]
I'm taking this issue out of our current milestone and pushing it back a little since I am still not able to reproduce the issue even with both the LDAP server and the PDNS auth server in two separate containers. (However, I'm not issueing `--network` commands to Docker).

@sooslaca (and perhaps @mind04) I'll wait for exact reproduction steps, otherwise it is going to take me too much time.

**Comment by sooslaca** [comment]
root@dockerhost:~/build/dh-pdns# cat Dockerfile
FROM alpine:3.6

# Execute build
COPY files/ /
RUN sh /build.sh

# Define start command
CMD [ "/run.sh" ]
--------------------------------------------------------------------------------------------------------------------------
root@dockerhost:~/build/dh-pdns# md5sum files/usr/src/pdns-4.1.0-rc1.tar.bz2
548017daae7907945c1b1ca4066f0e94  files/usr/src/pdns-4.1.0-rc1.tar.bz2
--------------------------------------------------------------------------------------------------------------------------
root@dockerhost:~/build/dh-pdns# cat files/build.sh
#!/usr/bin/env sh
set -ex

BUILD_DEPS="alpine-sdk boost-dev lua5.3-dev openldap-dev krb5-dev"
RUN_DEPS="inotify-tools openldap-clients lua5.3-libs libgcc libstdc++ boost-program_options libldap libsasl krb5-libs drill"

apk upgrade --no-cache

# install build dependencies
apk add --no-cache --virtual build-deps ${BUILD_DEPS}

# defaults from abuild.conf
export CFLAGS="-Os -fomit-frame-pointer"
export CXXFLAGS="$CFLAGS"
export CPPFLAGS="$CFLAGS"
export LDFLAGS="-Wl,--as-needed"

#export JOBS=2
export JOBS=$(getconf _NPROCESSORS_ONLN)
export MAKEFLAGS=-j$JOBS

USE_COLORS=1

# compile
mkdir -p /usr/src/pdns && cd /usr/src
tar jxf pdns-4.1.0-rc1.tar.bz2 -C /usr/src/pdns --strip-components=1
rm -f *.tar.bz2
cd /usr/src/pdns
./configure --prefix=/usr \
        --sysconfdir=/etc/pdns \
        --mandir=/usr/share/man \
        --infodir=/usr/share/info \
        --localstatedir=/var \
        --libdir=/usr/lib/pdns \
        --without-mysql \
        --with-modules="" \
        --with-dynmodules="bind ldap pipe" \
        --disable-static
make
make install

# remove build dependencies
apk del build-deps

# install run dependencies
apk add --no-cache ${RUN_DEPS}

# clean
rm -Rf /usr/src/pdns
rm -f /build.sh
--------------------------------------------------------------------------------------------------------------------------
root@dockerhost:~/build/dh-pdns# cat files/run.sh
#!/bin/sh

# main
/usr/sbin/pdns_server
--------------------------------------------------------------------------------------------------------------------------
--------------------------------------------------------------------------------------------------------------------------
--------------------------------------------------------------------------------------------------------------------------
root@dockerhost:~/build/dh-pdns# cat /data/pdns/pdns.conf
allow-axfr-ips=<someips>
daemon=no
launch=ldap
master=yes
ldap-host=ldap://ldap.internal/
ldap-basedn=dc=org
ldap-binddn=cn=admin,dc=org
ldap-secret=password
ldap-method=simple
ldap-timeout=50
local-port=55
loglevel=8
--------------------------------------------------------------------------------------------------------------------------
Start cmd:
docker run --init --network=internal --name pdns -p 10.10.10.10:53:55/udp -p 10.10.10.10:53:55 -v /data/pdns:/etc/pdns my/pdns

**Comment by Habbie** [comment]
```
59e27612 conn=1000 fd=15 ACCEPT from IP=172.17.0.3:45746 (IP=0.0.0.0:389)
59e27612 conn=1001 fd=16 ACCEPT from IP=172.17.0.3:45748 (IP=0.0.0.0:389)
59e27612 conn=1000 op=0 BIND dn="cn=admin,dc=org" method=128
59e27612 conn=1000 op=0 RESULT tag=97 err=49 text=
59e27612 conn=1002 fd=17 ACCEPT from IP=172.17.0.3:45750 (IP=0.0.0.0:389)
59e27612 conn=1002 op=0 BIND dn="cn=admin,dc=org" method=128
59e27612 conn=1002 op=0 RESULT tag=97 err=49 text=
59e27612 conn=1001 op=0 BIND dn="cn=admin,dc=org" method=128
59e27612 connection_input: conn=1001 deferring operation: binding
59e27612 conn=1001 op=0 RESULT tag=97 err=49 text=
59e27612 get_filter: conn 1001 unknown attribute type=PdnsDomainId (17)
59e27612 conn=1001 op=1 SRCH base="dc=org" scope=2 deref=3 filter="(&(sOARecord=*)(?PdnsDomainId=*))"
59e27612 conn=1001 op=1 SRCH attr=associatedDomain
59e27612 conn=1001 op=1 SEARCH RESULT tag=101 err=32 nentries=0 text=
59e27612 conn=1003 fd=18 ACCEPT from IP=172.17.0.3:45752 (IP=0.0.0.0:389)
59e27612 conn=1003 op=0 BIND dn="cn=admin,dc=org" method=128
59e27612 conn=1003 op=0 RESULT tag=97 err=49 text=
59e27612 conn=1004 fd=19 ACCEPT from IP=172.17.0.3:45754 (IP=0.0.0.0:389)
59e27612 conn=1004 op=0 BIND dn="cn=admin,dc=org" method=128
59e27612 conn=1004 op=0 RESULT tag=97 err=49 text=
59e2764e get_filter: conn 1001 unknown attribute type=PdnsDomainId (17)
59e2764e conn=1001 op=2 SRCH base="dc=org" scope=2 deref=3 filter="(&(sOARecord=*)(?PdnsDomainId=*))"
59e2764e conn=1001 op=2 SRCH attr=associatedDomain
59e2764e conn=1001 op=2 SEARCH RESULT tag=101 err=32 nentries=0 text=
```

```
Oct 14 20:39:45 Master/slave communicator launching
Oct 14 20:39:45 Creating backend connection for TCP
Oct 14 20:39:45 [LdapBackend] LDAP servers = ldap://172.17.0.2/
Oct 14 20:39:45 [LdapBackend] LDAP servers = ldap://172.17.0.2/
Oct 14 20:39:46 [LdapBackend] Ldap connection succeeded
Oct 14 20:39:46 About to create 3 backend threads for UDP
Oct 14 20:39:46 [LdapBackend] LDAP servers = ldap://172.17.0.2/
Oct 14 20:39:46 [LdapBackend] Ldap connection succeeded
Oct 14 20:39:46 [LdapBackend] Ldap connection succeeded
Oct 14 20:39:46 No master domains need notifications
Oct 14 20:39:46 [LdapBackend] LDAP servers = ldap://172.17.0.2/
Oct 14 20:39:46 [LdapBackend] Ldap connection succeeded
Oct 14 20:39:46 [LdapBackend] LDAP servers = ldap://172.17.0.2/
Oct 14 20:39:46 [LdapBackend] Ldap connection succeeded
Oct 14 20:39:46 Done launching threads, ready to distribute questions
```

no unbind, no crash, no nothing...

**Comment by mind04** [comment]
Looks like this is just a "slow query" and an ignored timeout setting.

https://github.com/PowerDNS/pdns/blob/master/modules/ldapbackend/powerldap.hh#L68
https://github.com/PowerDNS/pdns/blob/master/modules/ldapbackend/powerldap.hh#L73-L74

The default timeout value set here is not replaced with the value from the timeout setting.

**Comment by Habbie** [comment]
> The default timeout value set here is not replaced with the value from the timeout setting.

nice catch, that looks like it would explain things. @sooslaca can you patch the timeout values in 73/74 by hand and see if it helps?

**Comment by sooslaca** [comment]
I changed it to 50 secs
LINE 68
```
void bind( const string& ldapbinddn = "", const string& ldapsecret = "", int method = 
LDAP_AUTH_SIMPLE, int timeout = 50 );
```
LINE 73-74
```
bool getSearchEntry( int msgid, sentry_t& entry, bool dn = false, int timeout = 50 );
void getSearchResults( int msgid, sresult_t& result, bool dn = false, int timeout = 50 );
```
Now I get timeout after 50 secs.
```
Oct 15 20:26:19 dockerhost docker[19899]: Oct 15 20:26:19 db9877d53c9b daemon.notice pdns[2062]: [LdapBackend] Ldap connection succeeded
Oct 15 20:26:19 dockerhost docker[19899]: Oct 15 20:26:19 db9877d53c9b daemon.warn pdns[2062]: Done launching threads, ready to distribute questions
Oct 15 20:27:09 dockerhost docker[19899]: Oct 15 20:27:09 db9877d53c9b daemon.warn pdns[2062]: [LdapBackend] Unable to search LDAP directory: Timeout
Oct 15 20:27:09 dockerhost docker[19899]: Oct 15 20:27:09 db9877d53c9b daemon.notice pdns[2062]: [LdapBackend] Ldap connection closed
Oct 15 20:27:09 dockerhost docker[19899]: Oct 15 20:27:09 db9877d53c9b daemon.err pdns[2062]: Exiting because communicator thread died with error: LDAP server timeout
```

More findings: when it is at stage "Done launching threads", pdns looks like working. eg it responds on queries, etc... also validated it is actually reading LDAP as if I change an A record, pdns response reflects that.

Also note 4.0.4 works with default timeout (5 secs) and ldap container and pdns container are on the same host. Unlikely it needs more than 5 secs timeout.

Minor note - I know it is not related to this issue - accidentally while tested I also found axfr does not work. I get the SOA record but no other records. Story is the same: 4.0.4 with exact same values works.

One more test I can run is that I replace my ldap container with the osixia 1.1.9 written above as I use my custom one right now. Will report back

**Comment by sooslaca** [comment]
Ok I replaced my custom ldap with osixia and issue still exists

Oct 19 10:34:41 dockerhost docker[27035]: Oct 19 10:34:41 ab9df9ced4ff daemon.err pdns[11]: Master/slave communicator launching
Oct 19 10:34:41 dockerhost docker[27035]: Oct 19 10:34:41 ab9df9ced4ff daemon.info pdns[11]: [LdapBackend] LDAP servers = ldap://ldap2.internal/
Oct 19 10:34:41 dockerhost docker[27035]: Oct 19 10:34:41 ab9df9ced4ff daemon.err pdns[11]: Creating backend connection for TCP
Oct 19 10:34:41 dockerhost docker[27035]: Oct 19 10:34:41 ab9df9ced4ff daemon.info pdns[11]: [LdapBackend] LDAP servers = ldap://ldap2.internal/
Oct 19 10:34:41 dockerhost docker[27035]: Oct 19 10:34:41 ab9df9ced4ff daemon.notice pdns[11]: [LdapBackend] Ldap connection succeeded
Oct 19 10:34:41 dockerhost docker[27035]: Oct 19 10:34:41 ab9df9ced4ff daemon.notice pdns[11]: [LdapBackend] Ldap connection succeeded
Oct 19 10:34:41 dockerhost docker[27035]: Oct 19 10:34:41 ab9df9ced4ff daemon.warn pdns[11]: About to create 3 backend threads for UDP
Oct 19 10:34:41 dockerhost docker[27035]: Oct 19 10:34:41 ab9df9ced4ff daemon.info pdns[11]: [LdapBackend] LDAP servers = ldap://ldap2.internal/
Oct 19 10:34:41 dockerhost docker[27035]: Oct 19 10:34:41 ab9df9ced4ff daemon.notice pdns[11]: [LdapBackend] Ldap connection succeeded
Oct 19 10:34:41 dockerhost docker[27035]: Oct 19 10:34:41 ab9df9ced4ff daemon.info pdns[11]: [LdapBackend] LDAP servers = ldap://ldap2.internal/
Oct 19 10:34:41 dockerhost docker[27035]: Oct 19 10:34:41 ab9df9ced4ff daemon.notice pdns[11]: [LdapBackend] Ldap connection succeeded
Oct 19 10:34:41 dockerhost docker[27035]: Oct 19 10:34:41 ab9df9ced4ff daemon.info pdns[11]: [LdapBackend] LDAP servers = ldap://ldap2.internal/
Oct 19 10:34:41 dockerhost docker[27035]: Oct 19 10:34:41 ab9df9ced4ff daemon.notice pdns[11]: [LdapBackend] Ldap connection succeeded
Oct 19 10:34:41 dockerhost docker[27035]: Oct 19 10:34:41 ab9df9ced4ff daemon.warn pdns[11]: Done launching threads, ready to distribute questions

Oct 19 10:35:31 dockerhost docker[27035]: Oct 19 10:35:31 ab9df9ced4ff daemon.warn pdns[11]: [LdapBackend] Unable to search LDAP directory: Timeout
Oct 19 10:35:31 dockerhost docker[27035]: Oct 19 10:35:31 ab9df9ced4ff daemon.notice pdns[11]: [LdapBackend] Ldap connection closed
Oct 19 10:35:31 dockerhost docker[27035]: Oct 19 10:35:31 ab9df9ced4ff daemon.err pdns[11]: Exiting because communicator thread died with error: LDAP server timeout
Oct 19 10:35:32 dockerhost docker[27035]: Process not found, restarting

I'm out of ideas, moving back to 4.0.4 for now

**Comment by aerique** [comment]
@sooslaca Just to be complete sure: have you checked with @mind04's PR mentioned here https://github.com/PowerDNS/pdns/pull/5822 ?

**Comment by aerique** [comment]
Agh, this shouldn't have been automatically closed by #5822.

**Comment by sooslaca** [comment]
apologies for late answer. I did not know how to pull in #5822 so I waited for 4.1.0 that I assume has this PR in. 
Test results are still the same. 5 sec timeout
```
Dec 05 22:18:36 dockerhost docker[20290]: Dec  5 22:18:36 5fa46cd41d28 daemon.err pdns[370]: Master/slave communicator launching
....
Dec 05 22:18:36 dockerhost docker[20290]: Dec  5 22:18:36 5fa46cd41d28 daemon.err pdns[370]: Creating backend connection for TCP
....
Dec 05 22:18:36 dockerhost docker[20290]: Dec  5 22:18:36 5fa46cd41d28 daemon.warn pdns[370]: Done launching threads, ready to distribute questions
Dec 05 22:18:41 dockerhost docker[20290]: Dec  5 22:18:41 5fa46cd41d28 daemon.warn pdns[370]: [LdapBackend] Unable to search LDAP directory: Timeout
Dec 05 22:18:41 dockerhost docker[20290]: Dec  5 22:18:41 5fa46cd41d28 daemon.notice pdns[370]: [LdapBackend] Ldap connection closed
Dec 05 22:18:41 dockerhost docker[20290]: Dec  5 22:18:41 5fa46cd41d28 daemon.err pdns[370]: Exiting because communicator thread died with error: LDAP server timeout
```

The only this that we did not yet check is my LDAP entry. Am I using it right?

```
...snip...
PdnsDomainId: 1
PdnsDomainType: master
PdnsDomainMaster: <public IP of the master server where the other servers should axfr from?>
.../snip...
```

So I have this in my LDAP. Am I get it right:
with master=yes in pdns.conf what is happening is that server sends a dnsnotify to all servers mentioned in nSRecord attributes so those servers will pull zone via axfr?
Do I need to have an actual pdns slave to have this work? as I don't have one.

note: axfr still broken, I'll create another tkt for that. Broken in 4.0.5 as well, works in 4.0.4
